### PR TITLE
Database update & Serialization

### PR DESCRIPTION
**Pull request by blairyeung**
## DatabaseGataway update
I just updated the class `DatabaseGateway` and `DatabaseDataHandler`. The following functions are added:
- [+] Add 'saveToFile' to serialize the current data in the database to a file.
- [+] Add 'readFromFile' to read the serialized data from the file, into the database.
- Both of the methods does not accept any argument parameters, keep this in mind!
The name of the file is `data.ser`. It is placed (currently) under the root folder.
To call these methods, please use the singleton `DatabaseGateway` and call `databaseGateway.saveToFile()` or `databaseGateway.readFromFile()`, respectively. 

## Implementation of Serialization
I just implemented the above updates and now our project can serialize its data and store them in the file `data.ser`.
The program will save its data upon any update of entities. 
Also, we you starts the program, you will be able to read the entities/data from the previous run. If there is no previous records, the program will create a new `data.ser` for data storage.

### CODE REVIEW

**Comment by Teinble** [comment]
Also, I saw some methods' names are changed in this PR. Will these changes affect the other files?

**Comment by blairyeung** [comment]
> Also, I saw some methods' names are changed in this PR. Will these changes affect the other files?

The changes are only in the databasegateway and will not affect the other functionalities. Also, the changes are only made on the parameters (i.e, change String to Integer), not the signature itself.

**Review by JY3150 — CHANGES_REQUESTED** [review]
In your change to test files, if you change the correct verification code needed for registering a user to random, none of the test would pass. Change the random number string back to the string "DebugCode" should suffice.

**Review by JY3150 — APPROVED** [review]
The problems in test are fixed.

**Review by Teinble — APPROVED** [review]
I checked this pr, it works fine. I will approve

File: src/main/java/database/DatabaseGateway.java [hunk 1]
```diff
@@ -112,9 +116,10 @@ public Post getPost(String courseCode, String postTopic) {
      * @return bool
      */
     @Override
+
     public boolean postExists(String courseCode, String postTopic) {
         Course course = getCourse(courseCode);
-        List<Post> posts = course.getPosts();
+        ArrayList<Post> posts = course.getPosts();
```
[the remark this answers is not in the document]
Comment (blairyeung) [reply src/main/java/database/DatabaseGateway.java:122]: Solved

**Pull request merged**